src/main.cpp

```diff
@@ -1616,17 +1621,24 @@ bool AcceptToMemoryPool(CTxMemPool& pool, CValidationState &state, const CTransa
             return error("AcceptToMemoryPool: BUG! PLEASE REPORT THIS! ConnectInputs failed against MANDATORY but not STANDARD flags %s", hash.ToString());
         }
 
-        // Store transaction in memory
-        pool.addUnchecked(hash, entry, !IsInitialBlockDownload(Params()));
+        {
+            // We lock to prevent other threads from accessing the mempool between adding and evicting
+            LOCK(pool.cs);
+            
+            // Store transaction in memory
+            pool.addUnchecked(hash, entry, !IsInitialBlockDownload(Params()));
+
+            // Add memory address index
+            if (fAddressIndex) {
+                pool.addAddressIndex(entry, view);
+            }
 
-        // Add memory address index
-        if (fAddressIndex) {
-            pool.addAddressIndex(entry, view);
-        }
+            // insightexplorer: Add memory spent index
+            if (fSpentIndex) {
+                pool.addSpentIndex(entry, view);
+            }
 
-        // insightexplorer: Add memory spent index
-        if (fSpentIndex) {
-            pool.addSpentIndex(entry, view);
+            pool.EnsureSizeLimit();
```

Comment:
We may be able to simplify all this by merging the two insightexplorer calls and the call to `EnsureSizeLimit()` into `addUnchecked()`, which is called from only this one place. Then `addUnchecked` can acquire the lock and it wouldn't have to be done here. However, `addUnchecked()` is called from several places in test code, which may break if this function does these additional things. Probably best to leave as it is here for now at least. 

One small change you could make is to have the two insightexplorer functions and `EnsureSizeLimit()` assert that the lock is held, instead of (recursively) calling `LOCK(cs)`.

Reply:
I like this suggestion, but let's do this as follow up. As a note, this may also require adding a call to `CTxMempool::SetMempoolCostLimit(...)` in the tests which call `CTxMempool::addUnchecked(...)`.